batch.go

```diff
@@ -243,6 +245,19 @@ func (batch *Batch) readMessage(
 			// consumed or a batch whose connection is in an error state.
 			batch.err = dontExpectEOF(err)
 		case batch.msgs.remaining() == 0:
+			// Log compaction can create batches with 0 messages.
+			//
+			// If the "next offset" reaches the "originally requested offset"
+			// and we have 0 messages remaining, then there were 0 messages in
+			// the batch.
+			//
+			// We normally set the batch offset to the "next" batch offset upon
+			// reading a message but since there were no messages to read we
+			// update it now instead.
+			if batch.offset == batch.conn.offset {
+				batch.offset++
```

Comment:
Where a batch contains two compacted records at its tail advancing the offset by one will cause a subsequent read to retrieve the same batch (as we've not fully passed it). When that batch is processed a second time, it will read a valid record (no error will be returned) and our batch.offset will be moved backwards as we fall into the `case nil:` and the following is applied: `batch.offset = offset + 1`.

Reply:
One other option could be to only advance the `batch.offset` under the `err = nil` case where <= to the returned offset (meaning all `batch.offset++` increments in the `errShortRead` case will e preserved)

```
		if batch.offset <= offset {
			batch.offset = offset + 1
		}
```

Reply:
@leedavis81 If I understand correctly, we only enter the `errShortRead` state when we reach the end of a batch so advancing the offset by 1 should advance us to the next batch.

I think (not certain) that compacted records in the middle of a batch will only show up in the offset delta and won't cause a `errShortRead`.

I wonder if this new scenario is a different one? Are there any more details you could dig up about where the reader is getting stuck and what the kafka log segment looks like? I tried playing around with the unit test in this PR by increasing the segment size and also varying the duplicated messages but haven't been able to get stuck anywhere else yet.

Reply:
So the scenario you want is a batch with 3 records, where two are compacted at the end (obviously need to be differing keys). What happens here is the consumer reads record 1 OK, then moves onto record 2 where an errShortRead is triggered and you fall into the case where no records remain and the batch.offset is advanced (as per the fix in this PR).

As no messages reside in the batch, attempts to read batch+3 fail and a request is made to the broker for that offset. The broker then returns the same batch (as that's where the compacted entry resides) and the consumer begins to read that (still with batch.offset set to offset+3). The first read succeeds and returns a new offset with no error, falling into the err = nil case the batch.offset is then moved back to offset+1 and the cycle repeats.

Reply:
@nlsun this is a fix I used today to resolve the production issues we were experiencing.. https://github.com/leedavis81/kafka-go/pull/1/files

Reply:
last successfully read offset would be returned incremented by one after batch parsing.
not batch offset.

we only have a problem when there is compacted messages at the end of batch, leading us to have no new messages.
if we have no new messages - we have no new offset.

BUT if we increment offset manually - that case would be avoided.

Reply:
> atch.offset is advanced (as per the fix in this PR).

no. if we already read at least 1 message - then batch.offsed is not equal to batch.conn.offset.

we only advanse offset if there is no new messages in batch. and that incremented offset will be saved in batch.conn during batch.close() call.

thats why:
1) send Fetch req with offset 1
2) got batch with offset 1 and no more
3) set batch.offset to 1+1 and during `batch.close()` set `conn.offset = batch.offset`
4) send Fetch req with offset 2
5) got no new messages, leading to `batch.offset == batch.conn.offset` and `batch.offset` would be incremented by 1 to  3
6) during batch.close() call conn.offset will become 3 as well
7) send Fetch req with offset 3
8) got no new messages, move batch.offset to 4
9) PROFIT! =)

Reply:
It sounds to me like there's yet another condition that is causing a `errShortRead` that is currently unhandled that is not the "read 0 new records in a batch due to compacted records" being handled in this fix.

Are we getting an `errShortRead` for a reason besides reaching the end of a batch? My understanding is that compacted records don't actually show up as a message and are just reflected in the next non-compacted message having a higher offset delta

Reply:
Yeah that's my understanding too, where you have either an empty batch (record is compacted) or a _single_ trailing compacted record in a batch, then this fix will be fine.

There is however a scenario where you could have multiple compacted records at the end of the batch, in which this wouldn't solve as we would never increment past the last (or second compacted) entry (based on the description I detailed above).

Reply:
Ok, so I've got a heavily modified branch with a lot of print statements that has an improved test that tests varying lengths of consecutive compacted records. From reading the print statements I'm pretty confident that we can successfully cross multiple compacted records.

branch: https://github.com/segmentio/kafka-go/compare/fix-batch-offset-advance-nlsunsuggestion-multicompact-test?expand=1 and you can run the test with
```
go test -run TestReaderReadCompactedMessage -count 1 ./...
```

Reply:
@leedavis81 what do you think about tracking this additional bug in a separate github issue? That way we can merge this specific fix for those that need it.

My understanding is that the core issue is if something causes us to stop reading a batch then the `batch.offset` will be reset to whatever offset kafka sends back when starting a new connection. We are now handling the case where the end of a batch causes us to stop reading, but it seems like you are running into an additional issue that is causing the reader to exit as well.

It could be that we simply want to prevent the `batch.offset` from going backwards in which case the workaround in your branch would be best, which would be good to discuss in a separate issue

Reply:
Your understanding is spot on, and moving that to a seperate issue is perfectly fine. I do intend on testing your branch shortly, just haven't had the time to do it yet.

Reply:
So I was talking to someone about this and they got me thinking about `lastOffsetDelta` again, and I remembered that in the debug branch it was always `1` which I tried to change before but gave up before. This time I tried changing the `BatchSize` on the writer from `2` to `3` and bam it got stuck on what hopefully is what you're running into, will let you know when I fix this.

Reply:
@leedavis81 alright, hoping the million-th time's the charm eh? Could you try out this newly updated PR on your test case?

Reply:
i'm running this test now, should have some feedback shortly

Reply:
Hi @nlsun , I can confirm this patch resolved both instances where we had single and multiple compacted records at the tail of a batch. Nice work! 

This fix would close out https://github.com/segmentio/kafka-go/pull/709 (and a handful of other PRs by the looks of it)